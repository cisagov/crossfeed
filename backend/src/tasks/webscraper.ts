import {
  connectToDatabase,
  Domain,
  Organization,
  Scan,
  Service
} from '../models';
import { CommandOptions } from './ecs-client';
import { getLiveWebsites } from './helpers/getLiveWebsites';
import { spawn } from 'child_process';
import * as path from 'path';
import { writeFileSync } from 'fs';
import saveWebpagesToDb from './helpers/saveWebpagesToDb';
import * as readline from 'readline';
import PQueue from 'p-queue';
import { chunk } from 'lodash';
import { In } from 'typeorm';

const WEBSCRAPER_DIRECTORY = '/app/worker/webscraper';
const INPUT_PATH = path.join(WEBSCRAPER_DIRECTORY, 'domains.txt');
const WEBPAGE_DB_BATCH_LENGTH = process.env.IS_LOCAL ? 2 : 10;

// Sync this with backend/worker/webscraper/webscraper/items.py
export interface ScraperItem {
  url: string;
  status: number;
  domain_name: string;
  response_size: number;

  body: string;
  headers: { name: string; value: string }[];

  domain?: { id: string };
  discoveredBy?: { id: string };
}

export const handler = async (commandOptions: CommandOptions) => {
  const { chunkNumber, numChunks, organizationId, scanId } = commandOptions;

  await connectToDatabase();

  const scan = await Scan.findOne(
    { id: scanId },
    { relations: ['organizations'] }
  );

  // webscraper is a global scan, so organizationId is only specified for tests.
  // Otherwise, scan.organizations can be used for granular control of censys.
  const orgs = organizationId
    ? [organizationId]
    : scan?.organizations?.length
    ? scan?.organizations.map((org) => org.id)
    : undefined;

  const where = orgs?.length ? { id: In(orgs) } : {};
  const organizations = await Organization.find({ where, select: ['id'] });
  const organizationIds = (
    chunk(organizations, numChunks)[chunkNumber!] || []
  ).map((e) => e.id);
  console.log('Running webscraper on organizations ', organizationIds);

  const liveWebsites = await getLiveWebsites(undefined, organizationIds);
  const urls = liveWebsites.map((domain) => domain.url);
  console.log('input urls', urls);
  if (urls.length === 0) {
    console.log('no urls, returning');
    return;
  }

  writeFileSync(INPUT_PATH, urls.join('\n'));

  const liveWebsitesMap: { [x: string]: Domain } = {};
  for (const domain of liveWebsites) {
    liveWebsitesMap[domain.name] = domain;
  }
  let totalNumWebpages = 0;
  const queue = new PQueue({ concurrency: 1 });

  const scrapyProcess = spawn(
    'scrapy',
    ['crawl', 'main', '-a', `domains_file=${INPUT_PATH}`],
    {
      cwd: WEBSCRAPER_DIRECTORY,
      env: {
        ...process.env,
        HTTP_PROXY: process.env.GLOBAL_AGENT_HTTP_PROXY,
        HTTPS_PROXY: process.env.GLOBAL_AGENT_HTTP_PROXY
      },
      stdio: 'pipe'
    }
  );

  const readInterface = readline.createInterface({
    // Database output
    input: scrapyProcess.stdout
  });
  const readInterfaceStderr = readline.createInterface({
    // Scrapy logs
    input: scrapyProcess.stderr
  });

  await new Promise((resolve, reject) => {
    console.log('Going to save webpages to the database...');
    let scrapedWebpages: ScraperItem[] = [];
    readInterfaceStderr.on('line', (line) =>
      console.error(line?.substring(0, 999))
    );
    readInterface.on('line', async (line) => {
      if (!line?.trim() || line.indexOf('database_output: ') === -1) {
        console.log(line);
        return;
      }
      const item: ScraperItem = JSON.parse(
        line.slice(
          line.indexOf('database_output: ') + 'database_output: '.length
        )
      );
      const domain = liveWebsitesMap[item.domain_name];
      if (!domain) {
        console.error(
          `No corresponding domain found for item with domain_name ${item.domain_name}.`
        );
        return;
      }
      scrapedWebpages.push({
        ...item,
        domain: { id: domain.id },
        discoveredBy: { id: scanId }
      });
      if (scrapedWebpages.length >= WEBPAGE_DB_BATCH_LENGTH) {
        await queue.onIdle();
        queue.add(async () => {
          if (scrapedWebpages.length === 0) {
            return;
          }
          console.log(
            `Saving ${scrapedWebpages.length} webpages, starting with ${scrapedWebpages[0].url}...`
          );
          await saveWebpagesToDb(scrapedWebpages);
          totalNumWebpages += scrapedWebpages.length;
          scrapedWebpages = [];
        });
      }
    });

    readInterface.on('close', async () => {
      await queue.onIdle();
      if (scrapedWebpages.length > 0) {
        console.log(
          `Saving ${scrapedWebpages.length} webpages to the database...`
        );
        await saveWebpagesToDb(scrapedWebpages);
        totalNumWebpages += scrapedWebpages.length;
        scrapedWebpages = [];
      }
      resolve();
    });
    readInterface.on('SIGINT', reject);
    readInterface.on('SIGCONT', reject);
    readInterface.on('SIGTSTP', reject);
  });
  console.log(
    `Webscraper finished for ${liveWebsites.length} domains, saved ${totalNumWebpages} webpages in total`
  );
};
