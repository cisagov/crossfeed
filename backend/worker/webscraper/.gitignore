s3-data
out.jl
domains.txt