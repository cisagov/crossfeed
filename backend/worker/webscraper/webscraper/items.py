# Define here the models for your scraped items
#
# See documentation in:
# https://docs.scrapy.org/en/latest/topics/items.html

import scrapy
