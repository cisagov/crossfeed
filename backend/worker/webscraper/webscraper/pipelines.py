from scrapy.exceptions import DropItem
import json
import os
from io import BytesIO
from datetime import datetime


class ExportFilePipeline:
    """Prints file contents to the console."""

    def __init__(self, print=print):
        self.urls_seen = set()
        self.print = print

    def process_item(self, item, spider=None):
        if item["url"] in self.urls_seen:
            raise DropItem("Duplicate item found with url: %s" % item["url"])
        self.urls_seen.add(item["url"])
        self.print("database_output: " + json.dumps(item))
        return item
