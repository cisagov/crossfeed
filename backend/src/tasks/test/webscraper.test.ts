import { handler as webscraper } from '../webscraper';
import {
  connectToDatabase,
  Organization,
  Service,
  Domain,
  Webpage,
  Scan
} from '../../models';
import { spawn } from 'child_process';
import { writeFileSync } from 'fs';
import { Readable } from 'stream';
jest.mock('../es-client');

const updateWebpages = require('../es-client').updateWebpages as jest.Mock;

jest.mock('child_process', () => ({
  spawn: jest.fn().mockImplementationOnce(() => ({
    status: 0,
    stderr: Readable.from(['scrapy output']),
    stdout: Readable.from([
      `
line with no database output
database_output: {"body": "abc", "headers": [{"a": "b", "c": "d"}], "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "1de6816e1b0d07840082bed89f852b3f10688a5df6877a97460dbc474195d5dd", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/scans/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "6a2946030f804a281a0141397dbd948d7cae4698118bffd1c58e6d5f87480435", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/contributing/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "d8f190dfeaba948e31fc26e3ab7b7c774b1fbf1f6caac535e4837595eadf4795", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/usage/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"body": "abc", "headers": [{"a": "b", "c": "d"}], "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "226706ff585e907aa977323c404b9889f3b2e547d134060ed57fda2e2f1b9860", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/contributing/deployment/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "e1448fa789c02ddc90a37803150923359a4a21512e1737caec52be53ef3aa3b5", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/contributing/architecture/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "3091ca75bf2ee1e0bead7475adb2db8362f96b472d220fd0c29eaac639cdf37f", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/usage/customization/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
line with no database output {}
database_output: {"s3_key": "dd7e7e51a4a094e87ad88756842854c2d878ac55fb908035ddaf229c5568fa1a", "status": 200, "url": "https://docs.crossfeed.cyber.dhs.gov/usage/administration/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "afb6378d30bd1b43d86be5d1582d0e306ba6c9dd2c60f0dcbb1a3837b34cbe59", "status": 400, "url": "https://docs.crossfeed.cyber.dhs.gov/contributing/setup/", "domain_name": "docs.crossfeed.cyber.dhs.gov"}
database_output: {"s3_key": "afb6378d30bd1b43d86be5d1582d0e306ba6c9dd2c60f0dcbb1a3837b34cbe60", "status": 200, "url": "https://unrelated-url.cyber.dhs.gov/contributing/setup/", "domain_name": "unrelated-url.cyber.dhs.gov"}
`
    ])
  }))
}));

jest.mock('fs', () => ({
  writeFileSync: jest.fn()
}));

describe('webscraper', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  let organization;
  beforeEach(async () => {
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
  });
  test('basic test', async () => {
    const scan = await Scan.create({
      name: 'webscraper',
      arguments: {},
      frequency: 999
    }).save();
    const domain = await Domain.create({
      organization,
      name: 'docs.crossfeed.cyber.dhs.gov',
      ip: '0.0.0.0'
    }).save();
    const service = await Service.create({
      service: 'https',
      port: 443,
      domain
    }).save();
    await webscraper({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId',
      chunkNumber: 0,
      numChunks: 1
    });
    const webpages = await Webpage.find({
      where: { domain },
      relations: ['discoveredBy']
    });
    expect(
      (spawn as jest.Mock).mock.calls.map((e) => [e[0], e[1]])
    ).toMatchSnapshot();
    expect((writeFileSync as jest.Mock).mock.calls).toMatchSnapshot();
    expect(webpages.map((e) => e.url)).toMatchSnapshot();
    expect(webpages.map((e) => e.status)).toMatchSnapshot();
    expect(webpages.map((e) => e.headers)).toMatchSnapshot();
    expect(webpages.filter((e) => e.discoveredBy.id !== scan.id)).toEqual([]);

    expect(
      updateWebpages.mock.calls[0][0].map((e) => ({
        ...e,
        webpage_createdAt: e.webpage_createdAt ? true : false,
        webpage_updatedAt: e.webpage_updatedAt ? true : false,
        webpage_syncedAt: e.webpage_discoveredById ? true : false,
        webpage_discoveredById: e.webpage_discoveredById ? true : false,
        webpage_domainId: e.webpage_domainId ? true : false,
        webpage_lastSeen: e.webpage_lastSeen ? true : false,
        webpage_id: e.webpage_id ? true : false
      }))
    ).toMatchSnapshot();
  });
});
