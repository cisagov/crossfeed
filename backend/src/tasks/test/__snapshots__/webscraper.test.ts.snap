// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webscraper basic test 1`] = `
Array [
  Array [
    "scrapy",
    Array [
      "crawl",
      "main",
      "-a",
      "domains_file=/app/worker/webscraper/domains.txt",
    ],
  ],
]
`;

exports[`webscraper basic test 2`] = `
Array [
  Array [
    "/app/worker/webscraper/domains.txt",
    "https://docs.crossfeed.cyber.dhs.gov",
  ],
]
`;

exports[`webscraper basic test 3`] = `
Array [
  "https://docs.crossfeed.cyber.dhs.gov/scans/",
  "https://docs.crossfeed.cyber.dhs.gov/contributing/",
  "https://docs.crossfeed.cyber.dhs.gov/usage/",
  "https://docs.crossfeed.cyber.dhs.gov",
  "https://docs.crossfeed.cyber.dhs.gov/contributing/deployment/",
  "https://docs.crossfeed.cyber.dhs.gov/contributing/architecture/",
  "https://docs.crossfeed.cyber.dhs.gov/usage/customization/",
  "https://docs.crossfeed.cyber.dhs.gov/usage/administration/",
  "https://docs.crossfeed.cyber.dhs.gov/contributing/setup/",
]
`;

exports[`webscraper basic test 4`] = `
Array [
  "200",
  "200",
  "200",
  "200",
  "200",
  "200",
  "200",
  "200",
  "400",
]
`;

exports[`webscraper basic test 5`] = `
Array [
  Array [],
  Array [],
  Array [],
  Array [
    Object {
      "a": "b",
      "c": "d",
    },
  ],
  Array [],
  Array [],
  Array [],
  Array [],
  Array [],
]
`;

exports[`webscraper basic test 6`] = `
Array [
  Object {
    "webpage_body": "abc",
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [
      Object {
        "a": "b",
        "c": "d",
      },
    ],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/scans/",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/contributing/",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/usage/",
  },
  Object {
    "webpage_body": "abc",
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [
      Object {
        "a": "b",
        "c": "d",
      },
    ],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/contributing/deployment/",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/contributing/architecture/",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/usage/customization/",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "200",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/usage/administration/",
  },
  Object {
    "webpage_body": undefined,
    "webpage_createdAt": true,
    "webpage_discoveredById": true,
    "webpage_domainId": true,
    "webpage_headers": Array [],
    "webpage_id": true,
    "webpage_lastSeen": true,
    "webpage_responseSize": null,
    "webpage_status": "400",
    "webpage_syncedAt": true,
    "webpage_updatedAt": true,
    "webpage_url": "https://docs.crossfeed.cyber.dhs.gov/contributing/setup/",
  },
]
`;
